fusion:
  # Topics
  cloudTopic: "/os_cloud_node/points"
  imuTopic: "/imu/data"
  imgTopic: "/image_proc_resize/image"  
  odomTopic: "/odometry/cam"

  # frames
  lidarFrame: "fusion_base_link"
  imuFrame: "fusion_base_link"
  odometryFrame: "odom"
  camFrame: "fusion_cam"
  mapFrame: "map"

  # Extrinsic parameter between IMU and Camera.
  estimateExtrinsic: 0   # 0  Have an accurate extrinsic parameters. We will trust the following imu^R_cam, imu^T_cam, don't change it.
                          # 1  Have an initial guess about extrinsic parameters. We will optimize around your initial guess.
  combined:
    pointFeatureDistThres: 2.0 # [m]
    projectionErrorThres: 2.0 # [pixel]
    depthAssociateBinNumber: 5 # number of bin threshold to associate
    stackPointTime: 2.0
    lidarSkip: 2

  optimization:
    timeLag: 3.0

  imu:
    # IMU Settings
    imuAccNoise: 3.9939570888238808e-03
    imuGyrNoise: 1.5636343949698187e-03
    imuAccBiasN: 6.4356659353532566e-05
    imuGyrBiasN: 3.5640318696367613e-05
    imuGravity: 9.80511
    imuRPYWeight: 0.01
    # Extrinsics (IMU -> GPS)
    imu2gpsTrans: [0.144, -0.161, 0.769]

  camera:
    rotImu2Cam: [0, 0, 1,
              -1, 0, 0,
              0, -1, 0]   
    transImu2Cam: [0.4681, 0.0, 0.2539]
    # calib: "cam0_pinhole.yaml"
    # imageWidth: 1920
    # imageHeight: 1080
    # focalLength: 460.0
    # windowSize: 10
    # numOfF: 1000
    # initDepth: 5.0

    calib: "cam0_resize.yaml"
    imageWidth: 480
    imageHeight: 270
    focalLength: 230.0
    windowSize: 10
    numOfF: 1000
    initDepth: 5.0

    #optimization parameters
    solverTime: 0.04  # max solver itration time (ms), to guarantee real time
    numIterations: 8   # max solver itrations, to guarantee real time
    keyframeParallax: 10.0 # keyframe selection threshold (pixel)

    #feature traker paprameters
    maxCnt: 100            # max feature number in feature tracking
    minDist: 10            # min distance between two features 
    FThreshold: 1.0        # ransac threshold (pixel)
    showTrack: 1           # publish tracking image as topic
    flowBack: 1            # perform forward and backward optical flow to improve feature tracking accuracy

    #unsynchronization parameters
    estimateTD: 0                      # online estimate time offset between camera and imu
    TD: 0.0                             # initial value of time offset. unit: s. readed image clock + td = real image clock (IMU clock)


  laser:
    # GPS Settings
    useImuHeadingInitialization: true           # if using GPS data, set to "true"
    useGpsElevation: false                      # if GPS elevation is bad, set to "false"
    gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
    poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
    
    # Export settings
    savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
    savePCDDirectory: "/Downloads/LOAM/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

    # Sensor Settings
    sensor: ouster                            # lidar sensor type, either 'velodyne' or 'ouster'
    N_SCAN: 128                                  # number of lidar channel (i.e., 16, 32, 64, 128)
    HORIZON_SCAN: 1024                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048)
    downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
    lidarMinRange: 2.0                          # default: 1.0, minimum lidar range to be used
    lidarMaxRange: 100.0                       # default: 1000.0, maximum lidar range to be used

    # Extrinsics (lidar -> IMU)
    transLidar2Imu: [-0.3787, 0.0, -0.6945]
    rotLidar2Imu: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]
    # rpyLidar2Imu: [1,  0, 0,
    #                0, 1, 0,
    #                 0, 0, 1]

    # LOAM feature threshold
    edgeThreshold: 1.0
    surfThreshold: 0.1
    edgeFeatureMinValidNum: 10
    surfFeatureMinValidNum: 100

    # voxel filter paprams
    odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
    mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
    mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

    # robot motion constraint (in case you are using a 2D robot)
    z_tollerance: 1000                            # meters
    rotation_tollerance: 1000                     # radians

    # CPU Params
    numberOfCores: 12                              # number of cores for mapping optimization
    mappingProcessInterval: 0.15                  # seconds, regulate mapping frequency

    # Surrounding map
    surroundingkeyframeAddingDistThreshold: 0.5   # meters, regulate keyframe adding threshold
    surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold
    surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
    surroundingKeyframeSearchRadiusLarge: 50.0    # meters, within n meters scan-to-map optimization (when loop closure disabled)
    surroundingKeyframeSearchRadiusSmall: 50.0     # meters, within n meters for stacking pointcloud 
    surroundingKeyframeStackingThreshold: 100       # number of stacking keyframes

    # Loop closure
    loopClosureEnableFlag: true
    loopClosureFrequency: 1.0                     # Hz, regulate loop closure constraint add frequency
    surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
    historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
    historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
    historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
    historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

    # Visualization
    globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
    globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
    globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

